### PR TITLE
util/must: add `Handle` for convenient failure handling

### LINKED ISSUE #107418: util/must: convenience wrapper for making many assertions
See #106508.

It's common to execute many assertions at once, see e.g. `storage.assertMVCCIteratorInvariants()`. Returning errors from each individual assertion can get tedious and clutter the code. It would be convenient to run many assertions in a block, if we can do so efficiently.

Here's an example of how this might look:

```go
err := must.With(ctx, func(m *must.Must) {
  m.True(true, "yep")
  m.False(false, "again")
  value, err := someFunc()
  m.NoError(err, "someFunc failed")
  m.Equal(value, "foo", "someFunc returned invalid value")
})
```

This would be implemented by throwing panics on assertion failures, and recovering the panic in `must.With`. A specific panic value would be thrown, wrapping the assertion error, and only this value would be recovered -- other panics would be propagated.

`must.With` itself would handle the assertion failure like any other assertion failure, i.e. fatal or propagate.

Jira issue: CRDB-30031

### PR DESCRIPTION
**Pull request by erikgrinaker**
**util/must: add `Handle` for convenient failure handling**

This patch adds `Handle()`, a convenience function that avoids returning assertion errors when running many assertions. Instead, non-fatal assertion failures throw panics that are caught by `Handle()` and converted to errors. Example usage:

```go
// nolint:errcheck
err := must.Handle(ctx, func(ctx context.Context) {
  must.True(ctx, true, "not true")
  must.Equal(ctx, 1, 1, "not equal")
})
```

Inside a `Handle` closure, must assertions never return errors, so they can be ignored. However, the errcheck linter will complain about this and must be disabled.

Initially, an attempt was made to pass in a struct with infallible assertion methods, for a safer API. Unfortunately, methods can't use generics, nor can function types, so this appears to be one of the only ways to achieve this.

Resolves #107418.

**util/must: add `DisableFatalAssertions` for tests**

**storage: use `must` for MVCC iterator assertions**

Epic: none
Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 31 of 31 files at r3, 8 of 8 files at r4, 2 of 2 files at r5, 31 of 33 files at r8, 8 of 8 files at r9, 2 of 2 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[`pkg/util/must/must.go` line 271 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NaRvgTJBRKUra8QQmVq:-NaRvgTLA8fWasCfMb5R:b-hnb3l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66526f1ddca3593d3758cff66b1568d6375bdc1/pkg/util/must/must.go#L271)):*
> ```Go
> 
> // handlePanic is thrown and recovered via Handle().
> type handlePanic struct{ err error }
> ```

Better yet: make this an `error`, (with `func Unwrap() error { return h.err }`)
Then use `errors.As()` in your recover handler above, like this:

```go
if r := recover(); r != nil {
   e, ok := r.(error)
   if !ok { panic(r) }
   if unused := handlePanic{}; !errors.As(e, &unused) { panic(r) }
   err = e
}
```

Why this makes sense:
- the code under Handle can be complex and wrap its errors using errors.Wrap or other things.
- you don't want to discard these layers in the return value of Handle.

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`pkg/util/must/must.go` line 271 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NaRvgTJBRKUra8QQmVq:-NaRzG9v9Zq1KUjMgL74:b-b37vop) ([raw file](https://github.com/cockroachdb/cockroach/blob/a66526f1ddca3593d3758cff66b1568d6375bdc1/pkg/util/must/must.go#L271)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Better yet: make this an `error`, (with `func Unwrap() error { return h.err }`)
Then use `errors.As()` in your recover handler above, like this:

```go
if r := recover(); r != nil {
   e, ok := r.(error)
   if !ok { panic(r) }
   if unused := handlePanic{}; !errors.As(e, &unused) { panic(r) }
   err = e
}
```

Why this makes sense:
- the code under Handle can be complex and wrap its errors using errors.Wrap or other things.
- you don't want to discard these layers in the return value of Handle.
</blockquote></details>

Sure, if it fits in the mid-stack inlining budget.

Does the overall approach seem reasonable though? I'm a bit lukewarm on the whole `nolint:errcheck` thing and lack of API safety, but seems like the lesser of evils.

**Comment by knz** [comment]
IMHO the approach is sound. Also there's precedentr: that's what testing.T does.

**Comment by erikgrinaker** [comment]
Ok thanks, I'll wrap this up then.

**Review by RaduBerinde — COMMENTED** [review]
Why do the `must.XX` APIs return errors in the first place? I don't see much of a usecase for that..

E.g. not much benefit to do
```
if err := must.True(x, "msg"); err != nil {
  return err
}
```
instead of
```
if !x {
  return errors.AssertionFailedf("msg")
}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NacYVK7C4yKEDAUKbx1:b-4crmbq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @knz)

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, and @knz)

___
*[`pkg/storage/intent_interleaving_iter.go` line 1348 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NacZl7Y01wfwf8WU822:-NacZl7Y01wfwf8WU823:bw63r0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d0879ddc081c2cc49cc08cb61f91813109f18/pkg/storage/intent_interleaving_iter.go#L1348)):*
> ```Go
> 		// At least one of the iterators must be valid. The iterator's validity state
> 		// should match i.iterValid.
> 		must.True(ctx, iterValid || intentValid, "i.valid=%t but both iterators are invalid", i.valid)
> ```

This code looks much cleaner now, nice!

One thing to keep in mind though is that now all the formatting arguments get boxed into `interface{}` in all cases, so we will allocate a lot more - we don't want that in hot paths.

**Review by erikgrinaker — COMMENTED** [review]
> Why do the `must.XX` APIs return errors in the first place? I don't see much of a usecase for that..

A couple of reasons:

* We want to fatal on assertion failures in non-release builds, to make sure we notice them. We want to return assertion errors in release builds, to avoid outages over trigger-happy assertions, but still want them to notify Sentry and log an error with a stack trace at the failure site.

* For other assertions, e.g. `must.Equal(a, b)`, we can automatically include the values in the failure messages without the caller having to manually pass them in the format string, and can generate better output like diffs in the case of large values (e.g. large structs). We can also have less trivial assertions like `must.Contains()`, although probably nothing too complex since we want them to be reasonably cheap. This is all purely for convenience though, and we can drop it if that's the consensus, but I find myself preferring Testify over testing.T and figured the same would hold here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NaeYrM14xuFjcqgu3UM:bs3eyje)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @RaduBerinde)

___
*[`pkg/storage/intent_interleaving_iter.go` line 1348 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NacZl7Y01wfwf8WU822:-NaeaP1dEk4iqmSP72_i:bnoaf0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d0879ddc081c2cc49cc08cb61f91813109f18/pkg/storage/intent_interleaving_iter.go#L1348)):*
> One thing to keep in mind though is that now all the formatting arguments get boxed into `interface{}` in all cases, so we will allocate a lot more - we don't want that in hot paths.

Oof, yeah, good point. Not a big deal here specifically, because these are all gated on `must.Expensive` and compiled out in most builds, but could be a problem in general.

Given the assertions all inline, I was hoping the compiler would be clever enough to avoid the heap escape on the happy path, or that we could avoid it with something like the below, but alas.

```go
func True(ctx context.Context, v bool, format string, args ...interface{}) error {
	if v {
		return nil
	}
	f, a := format, args
	return failDepth(ctx, 1, f, a...)
}
```

That said, the current approach for assertions also has this problem, so I guess it isn't strictly worse than today, but it would be nice to avoid it. Consider e.g.:

```go
foo := "bar"
if foo != "bar" {
	log.Fatalf(ctx, "%s != bar", foo) // foo escapes
}
```

Both `must` or `log.Fatalf` can avoid it with something like:

```go
foo := "bar"
if foo != "bar" {
	foo := foo // escapes on unhappy path
	must.Fail(ctx, "%s != bar", foo)
}
```

But I agree that this is an unfortunate API footgun. Ideas for how to improve this?

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @RaduBerinde)

___
*[`pkg/storage/intent_interleaving_iter.go` line 1348 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NacZl7Y01wfwf8WU822:-NaehK196YPE5MXbi3_q:bpx2jxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d0879ddc081c2cc49cc08cb61f91813109f18/pkg/storage/intent_interleaving_iter.go#L1348)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> One thing to keep in mind though is that now all the formatting arguments get boxed into `interface{}` in all cases, so we will allocate a lot more - we don't want that in hot paths.

Oof, yeah, good point. Not a big deal here specifically, because these are all gated on `must.Expensive` and compiled out in most builds, but could be a problem in general.

Given the assertions all inline, I was hoping the compiler would be clever enough to avoid the heap escape on the happy path, or that we could avoid it with something like the below, but alas.

```go
func True(ctx context.Context, v bool, format string, args ...interface{}) error {
	if v {
		return nil
	}
	f, a := format, args
	return failDepth(ctx, 1, f, a...)
}
```

That said, the current approach for assertions also has this problem, so I guess it isn't strictly worse than today, but it would be nice to avoid it. Consider e.g.:

```go
foo := "bar"
if foo != "bar" {
	log.Fatalf(ctx, "%s != bar", foo) // foo escapes
}
```

Both `must` or `log.Fatalf` can avoid it with something like:

```go
foo := "bar"
if foo != "bar" {
	foo := foo // escapes on unhappy path
	must.Fail(ctx, "%s != bar", foo)
}
```

But I agree that this is an unfortunate API footgun. Ideas for how to improve this?
</blockquote></details>

This is only the case for format args though, with the typed convenience API we copy these internally on the unhappy path, e.g.:

```go
foo := "bar"
err := must.Equal(ctx, foo, "bar", "failed") // does not escape
if err != nil {
	return err
}
```

**Review by RaduBerinde — COMMENTED** [review]
> -   We want to fatal on assertion failures in non-release builds, to make sure we notice them. We want to return assertion errors in release builds, to avoid outages over trigger-happy assertions, but still want them to notify Sentry and log an error with a stack trace at the failure site.

Can't we just make sure that `AssertionFailedf` does these things? At least the SQL layer reports assertion errors to Sentry (via `sqltelemetry.RecordError`).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NagMeJw-p_k3v1S8kjI:bcl3s5d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, and @knz)

**Comment by RaduBerinde** [comment]
*[`pkg/storage/intent_interleaving_iter.go` line 1348 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NacZl7Y01wfwf8WU822:-NagOVRt33oj8jXTbc6s:bui6ff4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d0879ddc081c2cc49cc08cb61f91813109f18/pkg/storage/intent_interleaving_iter.go#L1348)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

```go
foo := "bar"
if foo != "bar" {
	log.Fatalf(ctx, "%s != bar", foo) // foo escapes
}
```

The code above does not allocate unless we get into the branch with Fatalf. There's a difference between escaping (which applies to pointers, which `foo` here is not) and boxing, which means allocating something just so that an `interface{}` can refer to it (in the string case `runtime.convTstring`).

https://godbolt.org/z/999bGEq5E
</blockquote></details>

Not "applies to pointers", I should have said "applies when passing a pointer to an object".

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @RaduBerinde)

___
*[`pkg/storage/intent_interleaving_iter.go` line 1348 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-NacZl7Y01wfwf8WU822:-Nah0eEX3EyjSvQ7fIy9:b-cq5zyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed2d0879ddc081c2cc49cc08cb61f91813109f18/pkg/storage/intent_interleaving_iter.go#L1348)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not "applies to pointers", I should have said "applies when passing a pointer to an object".
</blockquote></details>

Right, that makes sense, thanks for catching this. I'll have a closer look at this when I can, but I agree, this seems like a showstopper so we may have to do something like what you're suggesting.

File: pkg/util/must/must.go [hunk 1]
```diff
@@ -176,6 +182,19 @@ var (
 	OnFail func(ctx context.Context, err error)
 )
 
+// DisableFatalAssertions disables fatal assertions for the duration of a test,
+// re-enabling them at the end. This is useful for tests that check assertion
+// failures as errors.
+func DisableFatalAssertions(t testing.TB) {
```
Comment (miretskiy) [inline pkg/util/must/must.go:188]: This adds a testing.TB dependency to non-testing code.
Maybe have a small local interface instead?

Comment (erikgrinaker) [reply pkg/util/must/must.go:188]: Yeah, we already do that in a bunch of places (see e.g. `storage.DisableMetamorphicSimpleValueEncoding`) so I just yoloed it, but let's do better.

Comment (erikgrinaker) [reply pkg/util/must/must.go:188]: Looks like I'm to blame for `DisableMetamorphicSimpleValueEncoding ` too, via #81284, must've missed that it wasn't a test file. :( I'll clean it up tomorrow.

**Review by knz — COMMENTED** [review]
> Can't we just make sure that `AssertionFailedf` does these things? At least the SQL layer reports assertion errors to Sentry (via `sqltelemetry.RecordError`).

We could but it's not exactly trivial. The errors constructors do not have access to context.Context, so it's hard to include meaningful detail if we fail there. Access to the ctx is what I particularly like about the `must` API.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NahVxpQ3t4VollFvTFl:bg71gy2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @miretskiy, and @RaduBerinde)

**Review by knz — COMMENTED** [review]
> Can't we just make sure that `AssertionFailedf` does these things?

An alternative could be to replace all calls to `errors.AssertionFailedf` and `errors.AssertionErrorWithWrappedErrf` to some other API that includes the `ctx` as argument. Then lint against use of the raw API.

Thoughts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NahWTr74Y5oxzG9n8KH:bhzkqug)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @miretskiy, and @RaduBerinde)

**Review by erikgrinaker — COMMENTED** [review]
Yeah, it'd be a very small helper API, basically what `must.Fail()` does.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107790#-:-NajmbZBBC1w5cMinpge:boetem)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, @miretskiy, and @RaduBerinde)

**Comment by erikgrinaker** [comment]
Closing, as `must` was reverted in #108272.

**Pull request closed without merging**